src/harness/core.ts

```diff
@@ -0,0 +1,671 @@
+        forEach(callback: (value: V, key: K, collection: this) => void): void;
+        keys(): IterableIterator<K>;
+        values(): IterableIterator<V>;
+        entries(): IterableIterator<[K, V]>;
+    }
+
+    /**
+     * A collection of key/value pairs internally sorted by key.
+     */
+    export class KeyedCollection<K, V> implements ReadonlyKeyedCollection<K, V> {
+        private _comparer: (a: K, b: K) => number;
+        private _keys: K[] = [];
+        private _values: V[] = [];
+        private _order: number[] = [];
+        private _version = 0;
+        private _copyOnWrite = false;
+
+        constructor(comparer: (a: K, b: K) => number) {
+            this._comparer = comparer;
+        }
+
+        public get size() {
+            return this._keys.length;
+        }
+
+        public has(key: K) {
+            return binarySearch(this._keys, key, identity, this._comparer) >= 0;
+        }
+
+        public get(key: K) {
+            const index = binarySearch(this._keys, key, identity, this._comparer);
+            return index >= 0 ? this._values[index] : undefined;
+        }
+
+        public set(key: K, value: V) {
+            const index = binarySearch(this._keys, key, identity, this._comparer);
+            if (index >= 0) {
+                this._values[index] = value;
+            }
+            else {
+                this.writePreamble();
+                insertAt(this._keys, ~index, key);
+                insertAt(this._values, ~index, value);
+                insertAt(this._order, ~index, this._version);
+                this.writePostScript();
+            }
+            return this;
+        }
+
+        public delete(key: K) {
+            const index = binarySearch(this._keys, key, identity, this._comparer);
+            if (index >= 0) {
+                this.writePreamble();
+                removeAt(this._keys, index);
+                removeAt(this._values, index);
+                removeAt(this._order, index);
+                this.writePostScript();
+                return true;
+            }
+            return false;
+        }
+
+        public clear() {
+            if (this.size > 0) {
+                this.writePreamble();
+                this._keys.length = 0;
+                this._values.length = 0;
+                this._order.length = 0;
+                this.writePostScript();
+            }
+        }
+
+        public forEach(callback: (value: V, key: K, collection: this) => void) {
+            const keys = this._keys;
+            const values = this._values;
+            const order = this.getInsertionOrder();
+            const version = this._version;
+            this._copyOnWrite = true;
+            for (const index of order) {
+                callback(values[index], keys[index], this);
+            }
+            if (version === this._version) {
+                this._copyOnWrite = false;
+            }
+        }
+
+        public * keys() {

```

Comment:
Style nit: I'm a believer that the `*` should attach to the function name for generator methods (especially since without a modifier `* func() {}` just feels odd), rather than floating with whitespace on either side. We should have a lint rule to enforce style here, one way or another.

Reply:
I disagree. I believe the `*` should be separate from the identifier.